### PR TITLE
[RFC] net: lib: aws_fota: Add FOTA self-check

### PR DESCRIPTION
**Pull request by sigvartmh**
This adds a self check in the `aws_fota` callback when it connects where
it checks whether the update succeeded or not. Then will report
`SUCCEEDED` or `FAILED` with regards to the outcome in the AWS IoT Job.

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

**Comment by sigvartmh** [comment]
## Working proof-of-concept for doing self-check after a reboot

Missing pieces in the design is how to actually do the check for each firmware upgrade type:
* Modem
* Application
* MCUBoot

I'll try to include some implementations by tomorrow for working self-checks. Currently it only sets the pending update flag and checks if it has it marks it as good and report `SUCCESS` to the AWS IoT Job.

### Known issue:
Checking for a job is triggered twice due to something coming in on the `/update/accepted` this can easily be fixed by adding another state.

**Comment by sigvartmh** [comment]
@jayteemo if you want to help out reviewing/designing feel free to comment.

**Comment by jayteemo** [comment]
the three state variables (`execution_state` `fota_state` `pending_update `) are a bit confusing.
i see how they are used, but it would be nice if they could be condensed... maybe just by adding a state to `fota_state` ?
related: `pending_update` is set true after the job is updated, should it be set before?  why wait for the response?

**Comment by jayteemo** [comment]
i've not used the `settings` module before, but it looks good 👍

**Comment by sigvartmh** [comment]
> the three state variables (`execution_state` `fota_state` `pending_update `) are a bit confusing.
> i see how they are used, but it would be nice if they could be condensed... maybe just by adding a state to `fota_state` ?

I'm not married to the naming so suggestions are fine :) `nextState` for the `fota_state` maybe so it reflects the variable it sets in `statusDetails`?

Not sure what the other two should be called maybe `aws_job_status` for `execution_state`.  Suggestions for `pending_update` @jayteemo ?

> related: `pending_update` is set true after the job is updated, should it be set before? why wait for the response?

If the response fails we shouldn't try to apply the update as the job would be an invalid state and we will have the wrong  version number for the job execution meaning we won't be able to update it after we have rebooted.

**Comment by jayteemo** [comment]
> If the response fails we shouldn't try to apply the update as the job would be an invalid state and we will have the wrong version number for the job execution meaning we won't be able to update it after we have rebooted.

i see.  what conditions would cause the job document to be rejected? 
how would a network issue or device reset at this point impact the process?

**Comment by jayteemo** [comment]
> Suggestions for `pending_update` @jayteemo ?

can a state be added to `enum fota_status`?   something like `PENDING_REBOOT`? 
and then the `pending_update` flag can be removed?

**Comment by sigvartmh** [comment]
> i see. what conditions would cause the job document to be rejected?

Actually I don't know all the cases. The most common is updating the job document before an update has been accepted. Which causes a version number mismatch. Also malformed JSON in the update payload.

> how would a network issue or device reset at this point impact the process?

So if the document is rejected or not accepted before you reset or reconnect(as long as you don't store the pending update state) you would actually grab the new version number from the job document and be able to continue as normal.

If the device is able to reset before the update is accepted could potentially be an issue depending on if we put the handling of a pending update into the Fota state as suggested.

If we encode it in the FOTA state it would be hard to figure out a good scheme for a resumption of the job but maybe it can be done by having a partial state between. I think it is a good idea to embed it into the fota state but I'm trying to consider if it could cause any issues down the line.

File: subsys/dfu/src/dfu_target_mcuboot.c [hunk 1]
```diff
@@ -106,10 +106,38 @@ static int settings_set(const char *key, size_t len_rd,
...
+}
+
 bool dfu_target_mcuboot_identify(const void *const buf)
 {
 	/* MCUBoot headers starts with 4 byte magic word */
-	return *((const u32_t *)buf) == MCUBOOT_HEADER_MAGIC;
+	if (*((const u32_t *)buf) == MCUBOOT_HEADER_MAGIC) {
+	struct mcuboot_img_header header;
+	u8_t flash_area_id = PM_MCUBOOT_PRIMARY_ID;
+	int err = boot_read_bank_header(flash_area_id, &header,
```
Comment (sigvartmh) [inline subsys/dfu/src/dfu_target_mcuboot.c]: Move this out of identity function

Comment (sigvartmh) [reply subsys/dfu/src/dfu_target_mcuboot.c]: done.

**Comment by sigvartmh** [comment]
This has actually sort of working version checks now.

**Comment by carlescufi** [comment]
Note: This is an automated message.
@sigvartmh this Pull Request has not been updated for more than a month, please close it if not applicable anymore.
@hakonfam  @oyvindronningstad  please close it if the author is not available

**Pull request closed without merging**